src/Storages/PostgreSQL/StoragePostgreSQLReplica.h

```diff
@@ -0,0 +1,107 @@
+#endif
+
+#if USE_LIBPQXX
+#include "PostgreSQLReplicationHandler.h"
+#include "PostgreSQLReplicaSettings.h"
+
+#include <Parsers/IAST.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTColumnDeclaration.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterCreateQuery.h>
+#include <Interpreters/ExpressionAnalyzer.h>
+#include <ext/shared_ptr_helper.h>
+
+
+namespace DB
+{
+
+class StoragePostgreSQLReplica final : public ext::shared_ptr_helper<StoragePostgreSQLReplica>, public IStorage
+{
+    friend struct ext::shared_ptr_helper<StoragePostgreSQLReplica>;
+
+public:
+    StoragePostgreSQLReplica(
+        const StorageID & table_id_,
+        StoragePtr nested_storage_,
+        const Context & context_);
+
+    String getName() const override { return "PostgreSQLReplica"; }
+
+    void startup() override;
+    void shutdown() override;
+
+    NamesAndTypesList getVirtuals() const override;
+
+    Pipe read(
+        const Names & column_names,
+        const StorageMetadataPtr & metadata_snapshot,
+        SelectQueryInfo & query_info,
+        const Context & context,
+        QueryProcessingStage::Enum processed_stage,
+        size_t max_block_size,
+        unsigned num_streams) override;
+
+    /// Called right after shutdown() in case of drop query
+    void shutdownFinal();
+
+    void createNestedIfNeeded(const std::function<PostgreSQLTableStructure()> & fetch_table_structure);
+
+    /// Can be nullptr
+    StoragePtr tryGetNested();
+
+    /// Throw if impossible to get
+    StoragePtr getNested();
+
+    Context makeNestedTableContext() const;
+
+    void setNestedLoaded() { nested_loaded.store(true); }
+
+    bool isNestedLoaded() { return nested_loaded.load(); }
+
+    void dropNested();
+
+protected:
+    StoragePostgreSQLReplica(
+        const StorageID & table_id_,
+        const String & remote_database_name,
+        const String & remote_table_name,
+        const String & connection_str,
+        const StorageInMemoryMetadata & storage_metadata,
+        const Context & context_,
+        std::unique_ptr<PostgreSQLReplicaSettings> replication_settings_);
+
+private:
+    static std::shared_ptr<ASTColumnDeclaration> getMaterializedColumnsDeclaration(
+            const String name, const String type, UInt64 default_value);
+
+    ASTPtr getColumnDeclaration(const DataTypePtr & data_type) const;
+
+    ASTPtr getCreateNestedTableQuery(const std::function<PostgreSQLTableStructure()> & fetch_table_structure);
+
+    std::string getNestedTableName() const;
+
+    std::string remote_table_name;
+    std::shared_ptr<Context> global_context;
+
+    std::unique_ptr<PostgreSQLReplicaSettings> replication_settings;
+    std::unique_ptr<PostgreSQLReplicationHandler> replication_handler;
+
+    std::atomic<bool> nested_loaded = false;
```

Comment:
What is the purpose if this flag? I can assume, that you were trying to fix some race, but it's not clear if the race is really fixed or just hidden from tsan by atomic flag.

Reply:
Nested tables are created from replication handler in a separate thread, so database engine is created before. This is just a flag to determine whether nested table was created (will be set to true from replication handler). In storage class it is used in `read` method to forbid read attempts to nested table if it is not yet created.

Reply:
Fixed - it does not rely on this atomic anymore (not present in read method). There is a similar variable now, but it is used only inside database engine to maintain an invariant (and its value is changed only once in storage's life).